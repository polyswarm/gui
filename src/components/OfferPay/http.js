import validator from 'validator';
import secp256k1 from 'secp256k1';

class HttpOfferPay {
  constructor(url) {
    this.url = url;
  }

  getUrlAccount(account) {
    return '?account=' + account;
  }

  pay(key, offer, amount) {
    const url = this.url;
    return new Promise((resolve, reject) => {
      if (validator.isUUID(offer.guid, 4)) {
        resolve();
      } else {
        reject('Invalid GUID');
      }
    }).then(() => new Promise((resolve, reject) => {
      if (web3Utils.isAddress(offer.author)) {
        resolve();
      } else {
        reject();
      }
    })).then(() =>  new Promise((resolve, reject) => {
      if (offer.initial >= amount) {
        resolve();
      } else {
        reject();
      }
    }))
    .then(() => new Promise((resolve, reject) => {
      if (amount && amount > 0) {
        const offerState = [];
        offerState.push(0); // is close
        offerState.push(0); // sequence
        offerState.push(0); // ambassador address
        offerState.push(0); // expert address
        offerState.push(0); //  msig address
        offerState.push(new BigNumber(offer.initial).minus(amount)); // new balance in nectar ambassador
        offerState.push(amount); // balance in nectar expert
        offerState.push(0); // token address
        offerState.push(0); // A globally-unique identifer for the Listing.
        offerState.push(new BigNumber(offer.initial)); // The Offer Amount.
        offerState.push(0); // Cryptographic hash of the Artifact.
        offerState.push(0); // The URI of the Artifact.
        offerState.push(0); // Engagement Deadline
        offerState.push(0); // Assertion Deadline
        offerState.push(0); // has the expert made commitment
        offerState.push(0); // “malicious” or “benign”
        offerState.push(0); // Information derived during Assertion generation

        let depositState = marshallState(offerState);

        const sig = secp256k1.sign(depositState, key);
  
        const v = sig.recovery + 27;
        const r = sig.signature.slice(0, 32);
        const s = sig.signature.slice(32, 64);
  
          const body = JSON.stringify({
          state: depositState,
          v,
          r,
          s
          });
        resolve(body);
      } else {
        reject('No payment amount specified');
      }
    })).then((body) => {
      const account = this.getUrlAccount(offer.author);
      return fetch(url + '/offers' + offer.guid + '/sendmsg' + account, {
        headers: {
          'Content-Type': 'application/json'
        },
        method: 'post',
        body: body
      });
    }).then(response => {
      if (response.ok) {
        return response;
      }
      return new Promise(resolve => {
        resolve(response.json());
      }).then(json => {
        throw Error(json.message);
      });
    }).then(response => response.json())
      .then(body => body.result);
  }

  marshallState(inputs) {
    var m = this.getBytes(inputs[0]);

    for(var i=1; i<inputs.length;i++) {
      m += this.getBytes(inputs[i]).substr(2, this.getBytes(inputs[i]).length);
    }
    return m;
  }
}
export default HttpOfferPay;
