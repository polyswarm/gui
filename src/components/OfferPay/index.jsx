// Vendor imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Uuid from 'uuid/v4';
import BigNumber from 'bignumber.js';
// Offer imports
import AnimatedInput from '../AnimatedInput';
import Button from '../Button';
import Header from '../Header';
import ModalPassword from '../ModalPassword';
// Component imports
import strings from './strings';
import HttpOfferPay from './http';

class OfferPay extends Component {
  constructor(props) {
    super(props);
    this.state = {
      reward_error: null,
      reward: null,
      duration: null,
      duration_error: null,
      expert: null,
      expert_error: null,
    };

    this.payExpert = this.payExpert.bind(this);
    this.addMessage = this.addMessage.bind(this);
    this.onClickHandler = this.onClickHandler.bind(this);
    this.onRewardChanged = this.onRewardChanged.bind(this);
    this.addPayRequest = this.addPayRequest.bind(this);
    this.removePayRequest = this.removePayRequest.bind(this);
    this.validateFields = this.validateFields.bind(this);
    this.onWalletChangeHandler = this.onWalletChangeHandler.bind(this);
  }

  componentDidMount() {
    const { props: { url } } = this;
    this.http = new HttpOfferPay(url);
  }

  render() {
    const { state: { reward, reward_error } } = this;
    const { props: { offer, url, walletList, addRequest, removeRequest,
      requestsInProgress, onBackPressed } } = this;

    const address = walletList.findIndex((wallet) => wallet.address === offer.author);
    const wallet = walletList[address];
    const title = strings.title + offer.expert;
    return (
      <div className='OfferPay'>
        <Header title={title}
          requests={requestsInProgress}
          back={true}
          onBack={onBackPressed}
          address={wallet.address}
          nct={wallet.nct}
          eth={wallet.eth}/>
        <ModalPassword
          ref={modal => (this.modal = modal)}
          url={url}
          walletList={walletList}
          address={address}
          onWalletChange={this.onWalletChangeHandler}
          addRequest={addRequest}
          removeRequest={removeRequest}/>
        <div className='OfferPay-Content'>
          <div className='OfferPay-Center'>
            <h2>{strings.instructions}</h2>
            <AnimatedInput type='number'
              onChange={this.onRewardChanged}
              error={reward_error}
              placeholder={strings.reward}
              input_id='reward'/>
            <div className='OfferPay-Upload'>
              <Button
                disabled={!reward || reward_error}
                onClick={this.onClickHandler}>
                {strings.pay}
              </Button>
            </div>
          </div>
        </div>
      </div>
    );
  }

  addMessage(message) {
    const {props: {addMessage}} = this;
    if (addMessage) {
      addMessage(message);
    }
  }
  
  onClickHandler() {
    this.modal.open();
  }
  
  onRewardChanged(reward) {
    this.setState({reward: reward}, () => {
      this.validateFields();
    });
  }
  
  onWalletChangeHandler(didUnlock) {
    const { props: { onWalletChange } } = this;
    if (onWalletChange) {
      onWalletChange();
    }
    if (didUnlock) {
      return this.payExpert();
    } else {
      return null;
    }
  }

  payExpert() {
    const { state: {reward, reward_error}, props: { offer, onBackPressed } } = this;

    const rewardWei = new BigNumber(reward).times(new BigNumber('1000000000000000000'));

    const http = this.http;
    if (reward &&!reward_error) {
      const uuid = Uuid();
      this.addPayRequest(uuid);
      return new Promise(resolve => {
        if (onBackPressed) {
          onBackPressed();
        }
        resolve();
      })
        .then(() => http.pay(offer.guid, rewardWei.toString()))
        .then(result => this.addMessage(result))
        .catch(error => {
          let errorMessage;
          if (!error || !error.message || error.message.length === 0) {
            errorMessage = strings.error;
          } else {
            errorMessage = error.message;
          }
          this.setState({ error: errorMessage });

          //Update app
          const { props: { onWalletChange, onError } } = this;
          if (onWalletChange) {
            onWalletChange(false);
          }
          if (onError) {
            onError(errorMessage);
          }
        })
        .then(() => {
          this.removePayRequest(uuid);
        });
    } else {
      return null;
    }
  }

  addPayRequest(id) {
    const { addRequest } = this.props;
    if (addRequest) {
      addRequest(strings.requestPay, id);
    }
  }

  removePayRequest(id) {
    const { removeRequest } = this.props;
    if (removeRequest) {
      removeRequest(strings.requestPay, id);
    }
  }

  validateFields() {
    const {state: {reward}} = this;
    const {props: {last}} = this;
    const min = new BigNumber('0.0625');
    const lastPay = new BigNumber(last);
    if (reward && new BigNumber(reward).comparedTo(min) < 0 ) {
      this.setState({reward_error: 'Reward below 0.0625 minimum.'});
    } else if (reward && new BigNumber(reward).comparedTo(lastPay) <= 0 ) {
      this.setState({reward_error: `Reward must be higher than last payment of ${lastPay.toString()}.`});
    } else {
      this.setState({reward_error: null});
    }
  }
}

OfferPay.propTypes = {
  walletList: PropTypes.array,
  address: PropTypes.number,
  onError: PropTypes.func,
  onWalletChange: PropTypes.func,
  addOffer: PropTypes.func,
  addRequest: PropTypes.func,
  removeRequest: PropTypes.func,
  url: PropTypes.string,
  last: PropTypes.string,
};
export default OfferPay;
