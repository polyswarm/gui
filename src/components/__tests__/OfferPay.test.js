import React from 'react';
import {render} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import OfferPay from '../OfferPay';

it('renders without crashing', () => {
  const wrapper = render(<OfferPay />);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});