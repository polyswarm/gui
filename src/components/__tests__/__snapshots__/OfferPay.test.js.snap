// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="OfferPay"
>
  <header
    class="Header"
  >
    <div
      class="Header-Row"
    >
      <div
        class="Header-Title"
      >
        <img
          alt="PolySwarm logo"
          class="Header-Logo"
          src="../public/img/back-arrow.svg"
        >
          
        </img>
        <h3>
          Pay expert
        </h3>
        <span
          class="RequestSpinner"
        >
          
        </span>
      </div>
      <div
        class="Header-Actions"
      >
        
      </div>
    </div>
    <div
      class="Header-Address"
    >
      <p
        class="Header-Divided"
        style="font-size:7.0588235294117645vw"
      >
        author
      </p>
      <p
        class="Header-Divided"
        style="font-size:7.0588235294117645vw"
      >
        1 NCT 
      </p>
      <p
        style="font-size:7.0588235294117645vw"
      >
        1 ETH
      </p>
    </div>
  </header>
  <div
    class="OfferPay-Content"
  >
    <div
      class="OfferPay-Center"
    >
      <h2>
        Enter amount to pay
      </h2>
      <div
        class="AnimatedInput"
      >
        <div>
          
        </div>
        <input
          class="AnimatedInput-Input"
          id="reward"
          type="number"
          value=""
        >
          
        </input>
      </div>
      <div
        class="OfferPay-Upload"
      >
        <button
          class="Button"
          disabled=""
        >
          Pay
        </button>
      </div>
    </div>
  </div>
</div>
`;
